### ISSUE TITLE
subs is too slow for large mappings

### ISSUE THREAD

**Issue by asmeurer** [sympy/sympy#10697]
If you have a large mapping, subs iterates over each term in the mapping, even if `simultaneous=True`. Contrast this with `xreplace`, which does the substitution all at once.

```
In [18]: a = (x + 1)**2 + 1

In [19]: mapping = {Symbol('x%d' % i): i for i in range(10000)}

In [20]: %timeit a.subs(mapping)
1 loop, best of 3: 4.93 s per loop

In [21]: %timeit a.subs(mapping, simultaneous=True)
1 loop, best of 3: 8.04 s per loop

In [22]: %timeit a.xreplace(mapping)
100000 loops, best of 3: 11.5 µs per loop
```

`subs` should only traverse an expression once when `simultaneous=True` (side note: I agree with the `XXX` comment in the code, simultaneous=True should be the default for dictionaries, maybe even the default always, but that's another issue). Even when simultaneous=False, it may be faster to recursively apply subs recursively to the mapping (in reverse), remove duplicates, and then apply the mapping all at once to the expression. We'd have to test it, but I suspect it would be faster, since in the common case the expression will be large and the terms in the mapping will be small.

**Comment by Curious72** [comment]
@asmeurer  Can i have a look at this?

**Comment by moorepants** [comment]
FYI, we have subs benchmarks that could be useful and be expanded:

http://www.moorepants.info/misc/sympy-asv/#large_exprs.TimeLargeExpressionOperations.time_subs

**Comment by asmeurer** [comment]
@moorepants I think that may be testing a different situation. This issue is specifically about the speed for large mappings, because of the loop. It's quadratic when it should be linear. Notice that for my test case no actual substitution is even done.

@Curious72 you don't need to ask for permission to look at an issue. You are always free to do so. For this issue I will likely fix it myself if no one else does but I haven't don't any work on it yet.

**Comment by moorepants** [comment]
Yes, it is testing something different. I just wanted to note that we could add benchmarks for performance issues here: https://github.com/sympy/sympy_benchmarks.

**Comment by Curious72** [comment]
@asmeurer  Do we need to sort the list before substituting in the expression , i guess that is the reason for slow mappings!

**Comment by asmeurer** [comment]
@Curious72 your dictionary keys are all the same, `Symbol('x')`. So your final dictionary is just `{x: 9999}` (check it with `print(mapping)`). When you have a dictionary comprehension and it sees the same key multiple times it just uses the one last seen. That's why in my example I include `i` in the symbol name, so that the mapping dictionary has 10000 items.
